packager/packager.cc

```diff
@@ -1542,6 +1556,83 @@ unique_ptr<PackageInfoImpl> createAndPopulatePackageInfo(core::GlobalState &gs,
+    if (!gs.packageDB().getPackageInfo(pkgName).exists()) {
+        return;
+    }
+    auto &pkgInfo = PackageInfoImpl::from(gs.packageDB().getPackageInfo(pkgName));
+    metadata.index[pkgName] = metadata.nextIndex;
+    metadata.lowLink[pkgName] = metadata.nextIndex;
+    metadata.nextIndex++;
+    metadata.stack.push_back(pkgName);
+    metadata.onStack[pkgName] = true;
+    for (auto &i : pkgInfo.importedPackageNames) {
+        if (i.type == ImportType::Test) {
+            continue;
+        }
+        if (metadata.index[i.name.mangledName] == 0) {
+            // This is a tree edge (ie. a forward edge that we haven't visited yet).
+            strongConnect(gs, metadata, i.name.mangledName);
+            if (metadata.index[i.name.mangledName] == 0) {
+                // this might mean the other package doesn't exist, but that should have been caught already
+                continue;
+            }
+            // Since we can follow any number of tree edges for lowLink, the lowLink of child is valid for this package
+            // too.
+            metadata.lowLink[pkgName] = std::min(metadata.lowLink[pkgName], metadata.lowLink[i.name.mangledName]);
+        } else if (metadata.onStack[i.name.mangledName]) {
+            // This is a back edge (edge to ancestor) or cross edge (edge to a different subtree). Since we can only
+            // follow at most one back/cross edge, the best update we can make to lowlink of the current package is the
+            // child's index.
+            metadata.lowLink[pkgName] = std::min(metadata.lowLink[pkgName], metadata.index[i.name.mangledName]);
+        }
+        // If the child package is already visited and not on the stack, it's in a different SCC, so no update to the
+        // lowlink.
+    }
+
+    if (metadata.index[pkgName] == metadata.lowLink[pkgName]) {
+        // This is the root of an SCC. This means that all packages on the stack from this package to the top of the top
+        // of the stack are in the same SCC. Pop the stack until we reach the root of the SCC, and assign them the same
+        // SCC ID.
+        core::packages::MangledName poppedPkgName;
+        do {
+            poppedPkgName = metadata.stack.back();
+            metadata.stack.pop_back();
+            metadata.onStack[poppedPkgName] = false;
+            PackageInfoImpl &poppedPkgInfo = PackageInfoImpl::from(gs.packageDB().getPackageInfo(poppedPkgName));
+            poppedPkgInfo.sccID = metadata.nextSCCId;
+        } while (poppedPkgName != pkgName);
+        metadata.nextSCCId++;
+    }
+}
+
+// Tarjan's algorithm for finding strongly connected components
+void computeSCCs(core::GlobalState &gs) {
+    Timer timeit(gs.tracer(), "computeSCCs");
+    ComputeSCCsMetadata metadata;
+    auto allPackages = gs.packageDB().packages();
+    for (auto package : allPackages) {
+        if (metadata.index[package] == 0) {
+            strongConnect(gs, metadata, package);
+        }
+    }
```

Comment:
It would be really nice if we had a `PackageRef` type so that we could iterate the package storage directly. Indirecting through the `UnorderedMap` that stores all the package info would be great to avoid, and then you could use the `PackageRef` to index vectors in `ComputeSCCsMetadata` as well.

If there's time coming up, that might be a nice way to speed up all interactions with the package db!

Reply:
I'm working on this, see here

https://github.com/sorbet/sorbet/pull/7713/commits/f27bd47ddc110186a291f0d28eae7f083e5c3db9#diff-4672c6d5bb71f0accd68f16712224583ef645293957bca3af48c9737bd5e95c2R222